### PR TITLE
[System]: `MonoTlsStream` is now `IDisposable` and cleans up on error.

### LINKED ISSUE #8689: Memory leak in HttpWebRequest / HttpWebResponse when using a WebProxy
This is a follow-on of sorts to bug [#7356](https://github.com/mono/mono/issues/7356) . The test is essentially the same thing except through a WebProxy.

## Steps to Reproduce

1. Use Ubuntu 16.04. Install mono-devel, mono-utils. mono-profiler, whatever is necessary to have the compiler and profiler. I doubt this is specific to the Linux version, however.  You will also need a proxy server through which to make the web requests.  Replace **yourproxyip** and **yourproxyport** withthe proper values for your proxy server.
2. The program below simply loops over a set of popular websites and gets their homepages through a proxy server, using HttpWebRequest and HttpWebResponse. Compile and run it under the profiler, e.g.
```
mono --gc=sgen --profile=log:heapshot=50gc TestMemGrow.exe
```
3. Let it run for several hours (3-4, or more). After that, to view the profiler output enter:

```
mprof-report output.mlpd > out.txt
grep System.Byte out.txt
(grep -F "System.String" | grep -Fv "System.String[]") < out.txt
```

You should see  lists of allocations of System.Byte[] and System.String that grow increasingly larger. You can leave the program running and re-execute those commands, to see the continued growth of memory usage as the profiler adds more heap shots to its data file.  In my testing the heapshots came about 10 minutes apart.

I left in the GC calls from my prior bug, and this example writes the web page data to a file.  It also writes exceptions to a file exceptions.txt, because it is possible (I really don't know) that it is the exception handling leaking memory, so that information might be useful.

To execute this code you will need a proxy server through which you make the web requests.  Fill in your proxy server's IP and port in the appropriate place in the code.

**CODE:**

```C#
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Runtime;

namespace TestMemGrow
{
    class Program
    {
        static void Main()
        {
		List<string> websites = new List<string>() { "http://www.amazon.com", "http://www.google.com", "http://www.yahoo.com", "http://www.ebay.com", "http://www.overstock.com",
                "http://www.marketwatch.com", "http://www.bn.com", "http://www.newegg.com", "http://www.wsj.com", "http://www.arstechnica.com", "http://www.slashdot.org",
                "http://www.mediaite.com", "http://www.disqus.com", "http://www.twitter.com", "http://www.snap.com", "http://www.facebook.com", "http://www.usps.com",
                "http://www.ups.com", "http://www.techcrunch.com", "http://www.oracle.com", "http://www.java.com", "http://www.apple.com", "http://www.microsoft.com",
                "http://www.ibm.com", "http://www.dell.com", "http://www.asus.com", "http://www.gigabyte.com", "http://www.intel.com", "http://www.crucial.com",
                "http://www.westerndigital.com", "http://www.samsung.com", "http://www.sandisk.com", "http://www.brother.com", "http://www.hp.com",
                "http://www.msn.com", "http://www.disney.com", "http://www.nintendo.com", "http://www.twitter.com", "http://www.youtube.com",
                "http://www.instagram.com", "http://www.linkedin.com", "http://www.wordpress.org", "http://www.pinterest.com", "http://www.wikipedia.org",
                "http://www.blogspot.com", "http://www.adobe.com", "http://www.tumblr.com", "http://www.vimeo.com", "http://www.flickr.com", "http://www.godaddy.com",
                "http://www.buydomains.com", "http://www.reddit.com", "http://www.w3.org","http://www.nytimes.com", "http://www.statcounter.com",
                "http://www.weebly.com","http://www.blogger.com","http://www.github.com", "http://www.jimdo.com", "http://www.myspace.com",
                "http://www.mozilla.org", "http://www.gravatar.com", "http://www.theguardian.com", "http://www.bluehost.com", "http://www.cnn.com", "http://www.foxnews.com",
                "http://www.msnbc.com", "http://www.wix.com", "http://www.paypal.com","http://www.stumbleupon.com", "http://www.digg.com","http://www.huffingtonpost.com",
                "http://www.feedburner.com", "http://www.imdb.com","http://www.yelp.com","http://www.dropbox.com", "http://www.baidu.com","http://www.washingtonpost.com",
                "http://www.slideshare.net","http://www.etsy.com","http://www.telegraph.co.uk", "http://www.about.com", "http://www.bing.com", "http://www.latimes.com",
                "http://www.tripadvisor.com","http://www.opera.com", "http://www.live.com", "http://www.wired.com", "http://www.bandcamp.com"};
			StreamWriter sw = new StreamWriter("exceptions.txt");

            while (true)
            {
                foreach (string website in websites)
                {
                    try
                    {
                        GCSettings.LargeObjectHeapCompactionMode = GCLargeObjectHeapCompactionMode.CompactOnce;
                        GC.Collect();
			Console.WriteLine(DateTime.Now.ToString("HH:mm:ss") + " Getting " + website);
                        HttpWebRequest request = (HttpWebRequest)WebRequest.Create(website);
			WebProxy myProxy = new WebProxy("yourproxyip", yourproxyport);
			request.Proxy = myProxy;
                        HttpWebResponse response = request.GetResponse() as HttpWebResponse;
                        StreamReader reader = new StreamReader(response.GetResponseStream());
                        File.WriteAllText("florb.txt", reader.ReadToEnd());
                        reader.Close();
                        response.Close();
                    }
                    catch (Exception ex)
                    {
			string exString = ex.ToString();
			exString = exString.Substring(0, Math.Min(150, exString.Length));
                        Console.WriteLine("Caught exception . . . " + exString);
			sw.WriteLine("EXCEPTION\r\n" + ex.ToString() + "\r\n");
			sw.Flush();
                    }
                }
            }
        }
    }
}
```

### Current Behavior

Memory allocations of System.Byte[] and System.String increase apparently without limit.

### Expected Behavior

The garbage collector would clean up these allocations.

## On which platforms did you notice this

[ ] macOS
[X] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 5.12.0.226 (tarball Thu May  3 09:48:32 UTC 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
```

## Other Matters

I am not at all certain whether this leak is caused by the use of a WebProxy itself, or by the increased number of exceptions, particularly WebExceptions, that occur when using a WebProxy to make the requests.

### PR DESCRIPTION
**Pull request by baulig**
When the TLS Handshake fails in `MonoTlsStream.CreateStream()`, then we must dispose
the `SslStream` to avoid a memory leak.

In addition to this, `MonoTlsStream` now implements `IDisposable` and is disposes
in `WebConnection.CloseSocket()`.

Fixes #8689

### CODE REVIEW

File: mcs/class/System/Mono.Net.Security/MonoTlsStream.cs [hunk 1]
```diff
@@ -154,5 +156,21 @@ internal async Task<Stream> CreateStream (WebConnectionTunnel tunnel, Cancellati
 			throw new PlatformNotSupportedException (EXCEPTION_MESSAGE);
 #endif
 		}
+
+		void Dispose (bool disposing)
+		{
+			if (Interlocked.CompareExchange (ref disposed, 1, 0) != 0)
```
Comment (akoeplinger) [inline mcs/class/System/Mono.Net.Security/MonoTlsStream.cs]: do we expect this to be called by multiple threads or why are you doing this vs. a simple `if (disposed) return;` check?

File: mcs/class/System/Mono.Net.Security/MonoTlsStream.cs [hunk 2]
```diff
@@ -136,6 +137,7 @@ internal async Task<Stream> CreateStream (WebConnectionTunnel tunnel, Cancellati
 					request.ServicePoint.UpdateClientCertificate (sslStream.InternalLocalCertificate);
 				else {
 					request.ServicePoint.UpdateClientCertificate (null);
+					sslStream?.Dispose ();
```
Comment (akoeplinger) [inline mcs/class/System/Mono.Net.Security/MonoTlsStream.cs]: how can sslStream be null here?

**Comment by nealef** [comment]
That will also fix the readBuffer and writeBuffer objects? (And were they supposed to be different sizes? 16834 looks like a typo)

File: mcs/class/System/Mono.Net.Security/MonoTlsStream.cs (hunk 1 above)
Comment (baulig) [reply mcs/class/System/Mono.Net.Security/MonoTlsStream.cs]: I'm using this as a pattern in the other related files - does it hurt to have the interlocked here?

File: mcs/class/System/Mono.Net.Security/MonoTlsStream.cs [hunk 3]
```diff
@@ -154,5 +156,21 @@ internal async Task<Stream> CreateStream (WebConnectionTunnel tunnel, Cancellati
...
+
+			if (sslStream != null) {
+				sslStream.Dispose ();
+				sslStream = null;
+			}
+		}
+
+		public void Dispose ()
+		{
+			Dispose (true);
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply mcs/class/System/Mono.Net.Security/MonoTlsStream.cs]: That's not needed when there is no finalizer but at the same time we don't need the argument either

**Comment by baulig** [comment]
@nealef I believe the 16834 was a type, just fixed that in the code.
@marek-safar @akoeplinger I also cleaned up the Dispose() a bit.

**Comment by nealef** [comment]
Thanks!


@nealef<https://github.com/nealef> I believe the 16834 was a type, just fixed that in the code.
@marek-safar<https://github.com/marek-safar> @akoeplinger<https://github.com/akoeplinger> I also cleaned up the Dispose() a bit.

File: mcs/class/System/Mono.Net.Security/MonoTlsStream.cs [hunk 4]
```diff
@@ -73,6 +73,7 @@ class MonoTlsStream
 #endif
 
 		WebExceptionStatus status;
+		int disposed;
```
Comment (marek-safar) [inline mcs/class/System/Mono.Net.Security/MonoTlsStream.cs]: This is not used anywhere

**Comment by akoeplinger** [comment]
how far back should we backport this?

**Comment by nealef** [comment]
2018-06 please.


-------- Original message --------

**Comment by nealef** [comment]
BTW, I did my testing against 2018-06 and it worked as expected.

**Comment by nealef** [comment]
Individually this fix and the ChainStatus fix appear to work. Now that both fixes are part of 2018-06 I cloned and checked out this branch and rebuilt. Now when I run the ChainStatus test case things break:

```
System.Net.WebException: A call to SSPI failed, see inner exception. ---> System.Security.Authentication.AuthenticationException: A call to SSPI failed, see inner exception. ---> System.InvalidOperationException: Operation is not valid due to the current state of the object.
  at Mono.Net.Security.MobileAuthenticatedStream.InnerRead (System.Boolean sync, System.Int32 requestedSize, System.Threading.CancellationToken cancellationToken) [0x00055] in <41884af155f44b798e04496ffbf2e844>:0 
  at Mono.Net.Security.AsyncProtocolRequest.InnerRead (System.Threading.CancellationToken cancellationToken) [0x000ac] in <41884af155f44b798e04496ffbf2e844>:0 
  at Mono.Net.Security.AsyncProtocolRequest.ProcessOperation (System.Threading.CancellationToken cancellationToken) [0x00093] in <41884af155f44b798e04496ffbf2e844>:0 
  at Mono.Net.Security.AsyncProtocolRequest.StartOperation (System.Threading.CancellationToken cancellationToken) [0x0008b] in <41884af155f44b798e04496ffbf2e844>:0 
   --- End of inner exception stack trace ---
  at Mono.Net.Security.MobileAuthenticatedStream.StartOperation (Mono.Net.Security.MobileAuthenticatedStream+OperationType type, Mono.Net.Security.AsyncProtocolRequest asyncRequest, System.Threading.CancellationToken cancellationToken) [0x0024b] in <41884af155f44b798e04496ffbf2e844>:0 
  at System.Net.WebResponseStream.InitReadAsync (System.Threading.CancellationToken cancellationToken) [0x000db] in <41884af155f44b798e04496ffbf2e844>:0 
  at System.Net.WebOperation.Run () [0x001d9] in <41884af155f44b798e04496ffbf2e844>:0 
  at System.Net.WebCompletionSource`1[T].WaitForCompletion () [0x00094] in <41884af155f44b798e04496ffbf2e844>:0 
  at System.Net.HttpWebRequest.MyGetResponseAsync (System.Threading.CancellationToken cancellationToken) [0x0027f] in <41884af155f44b798e04496ffbf2e844>:0 
   --- End of inner exception stack trace ---
  at System.Net.HttpWebRequest.RunWithTimeoutWorker[T] (System.Threading.Tasks.Task`1[TResult] workerTask, System.Int32 timeout, System.Action abort, System.Func`1[TResult] aborted, System.Threading.CancellationTokenSource cts) [0x000c5] in <41884af155f44b798e04496ffbf2e844>:0 
  at System.Net.HttpWebRequest.GetResponse () [0x00016] in <41884af155f44b798e04496ffbf2e844>:0 
  at ConsoleApplication3.TestHttpsRequest (System.String url, System.Boolean addCert) [0x0004e] in <d81caa8705284c178c7e0ddab8649693>:0 
  at ConsoleApplication3.TestCase4 () [0x00023] in <d81caa8705284c178c7e0ddab8649693>:0 
```
It's as if the initial certificate error is noticed but not that our callback is giving the go ahead for the session to proceed but the rug has been pulled out from under it.

**Comment by nealef** [comment]
I note that this problem has been reported elsewhere so I am not sure it is these fixes that are responsible. I did some MONO_TLS_DEBUG tracing and saw the critical piece is:
```
MobileAuthenticatedStream(1): AsyncReadRequest(1:75): ProcessOperation - read inner: 10917
MobileAuthenticatedStream(1): InnerRead: False 0 5515 10869 10917
Remaining: 10869 Requsted: 10917 Min: 10869
MobileAuthenticatedStream(1): InnerRead done: 10869 10869 - 10869
MobileAuthenticatedStream(1): InnerRead done: [BufferOffsetSize: 0 16384] - 10869 10869
MobileAuthenticatedStream(1): AsyncReadRequest(1:76): ProcessOperation - read inner done: 10917 - 10869
MobileAuthenticatedStream(1): AsyncReadRequest(1:77): ProcessOperation - read inner: 48
MobileAuthenticatedStream(1): InnerRead: False 0 16384 0 48
Remaining: 0 Requested: 48 Min: 0
```
Is seems we are told we need to read 10917 bytes but our read only returns 10869 so we go back and try and read the remaining 48 bytes but the readBuffer.Remaining value is 0 and the logic is:
```
                         var len = System.Math.Min (readBuffer.Remaining, requestedSize);
                        if (len == 0)
                                throw new InvalidOperationException ();
```
(This is in ```Mono.Net.Security/MobileAuthenticatedStream.cs```). The trace above shows the values and result of the System.Math.Min() function.

I think I found the problem but I don't know why. In ```MobileAuthenticatedStream.cs``` we used to set readBuffer to be of size 16834. In a recent push it was set to 16384 to match writeBuffer thinking it was simply a typo. If I return it to 16834 everything works again!

```
MobileAuthenticatedStream(1): AsyncReadRequest(1:75): ProcessOperation - read inner: 10917
MobileAuthenticatedStream(1): InnerRead: False 0 5515 11319 10917
Remaining: 11319 Requsted: 10917 Min: 10917
MobileAuthenticatedStream(1): InnerRead done: 11319 10917 - 10917
MobileAuthenticatedStream(1): InnerRead done: [BufferOffsetSize: 0 16432] - 10917 10917
MobileAuthenticatedStream(1): AsyncReadRequest(1:76): ProcessOperation - read inner done: 10917 - 10917
MobileAuthenticatedStream(1): AsyncReadRequest(1:77): ProcessOperation run: Continue
```

**Comment by nealef** [comment]
Referencing mono/mono#10371

**Comment by nealef** [comment]
It appears that during the exchange we are requesting 16432 bytes but the buffer only holds 16384 so there is still 48 remaining. However, before we go for that remaining 48 we ask for 10917 but only get 10869. I put a Console.Error.WriteLine() in the get_Remaining property and the other message comes just before the System.Math.Min() in InnerRead.
```
Remaining: 16384 Requested: 16432 Min: 16384                          
Length: 16384 Offset: 0 Size: 0                                      
Length: 16384 Offset: 0 Size: 5515                                   
Length: 16384 Offset: 0 Size: 5515                                   
Length: 16384 Offset: 0 Size: 5515                                   
Length: 16384 Offset: 0 Size: 5515                                   
Remaining: 10869 Requested: 10917 Min: 10869                          
Length: 16384 Offset: 0 Size: 5515                                   
Length: 16384 Offset: 0 Size: 16384                                  
Length: 16384 Offset: 0 Size: 16384                                  
Length: 16384 Offset: 0 Size: 16384                                  
Length: 16384 Offset: 0 Size: 16384                                  
Remaining: 0 Requested: 48 Min: 0
```
When the buffer is 16834 we are able to satisfy the first of those requests and proceed on:
```
Remaining: 16834 Requested: 16432 Min: 16432
Length: 16834 Offset: 0 Size: 0
Length: 16834 Offset: 0 Size: 5515
Length: 16834 Offset: 0 Size: 5515
Length: 16834 Offset: 0 Size: 5515
Length: 16834 Offset: 0 Size: 5515
Remaining: 11319 Requested: 10917 Min: 10917
Length: 16834 Offset: 0 Size: 5515
Length: 16834 Offset: 0 Size: 13795
Length: 16834 Offset: 0 Size: 13795
Length: 16834 Offset: 0 Size: 13795
Length: 16834 Offset: 0 Size: 13795
```

**Comment by nealef** [comment]
If ret < requested then should multiple reads be done until that request is fulfilled?

**Pull request merged**